### PR TITLE
Fixed #27482 -- Doc'd example of Case() w/ filter

### PR DESCRIPTION
**Pull request by joncasdam**
There wasn't in documentation an example of how to use Case() with filter() method. This PR adds that with a correspondent test for it.

https://code.djangoproject.com/ticket/27482

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, test this please.

File: docs/ref/models/conditional-expressions.txt [hunk 1]
```diff
@@ -141,6 +141,17 @@ the ``Client`` has been with us, we could do so using lookups::
     both Jane Doe and Jack Black. This works just like an :keyword:`if` ...
     :keyword:`elif` ... :keyword:`else` statement in ``Python``.
 
+``Case()`` also works with ``filter()`` to fetch objects::
+
+    >>> a_month_ago = date.today() - timedelta(days=30)
```
Comment (timgraham) [inline docs/ref/models/conditional-expressions.txt]: You didn't use `a_month_ago`. Is it a realistic query if both conditions use `a_year_ago`?

File: docs/ref/models/conditional-expressions.txt [hunk 2]
```diff
@@ -141,6 +141,17 @@ the ``Client`` has been with us, we could do so using lookups::
...
+    >>> a_year_ago = date.today() - timedelta(days=365)
+    >>> # Filter Client by registration date based on account type
+    >>> Client.objects.filter(registered_on__gte=Case(
+    ...     When(Q(account_type=Client.PLATINUM), then=a_year_ago),
+    ...     When(Q(account_type=Client.GOLD), then=a_year_ago))
```
Comment (timgraham) [inline docs/ref/models/conditional-expressions.txt]: put the closing ) on the next line

File: tests/expressions_case/tests.py [hunk 3]
```diff
@@ -1289,3 +1289,15 @@ def test_conditional_aggregation_example(self):
             ),
             {'regular': 2, 'gold': 1, 'platinum': 3}
         )
+
+    def test_example_with_filter(self):
+        a_year_ago = date.today() - timedelta(days=365)
+
```
Comment (timgraham) [inline tests/expressions_case/tests.py]: chop blank line

File: tests/expressions_case/tests.py [hunk 4]
```diff
@@ -1289,3 +1289,15 @@ def test_conditional_aggregation_example(self):
...
+        self.assertQuerysetEqual(
+            Client.objects.filter(registered_on__gte=Case(
+                When(Q(account_type=Client.PLATINUM), then=a_year_ago),
+                When(Q(account_type=Client.GOLD), then=a_year_ago))
```
Comment (timgraham) [inline tests/expressions_case/tests.py]: put closing ) on the next line

File: docs/ref/models/conditional-expressions.txt (hunk 2 above)
Comment (joncasdam) [inline docs/ref/models/conditional-expressions.txt]: ok

File: tests/expressions_case/tests.py (hunk 3 above)
Comment (joncasdam) [inline tests/expressions_case/tests.py]: ok

File: tests/expressions_case/tests.py (hunk 4 above)
Comment (joncasdam) [inline tests/expressions_case/tests.py]: ok

**Comment by joncasdam** [comment]
@timgraham changes amended

**Comment by timgraham** [comment]
Sorry, I guess my earlier comment wasn't clear. What I meant is that I'm not sure this example shows a benefit of `Case` compared to what seems like a simpler query: `Client.objects.filter(registered_on__gte=a_year_ago, account_type__in=(Client.PLATINUM, Client.GOLD)`. Are they logically the same or is there a subtle difference? If so, it would be nice to state that a bit more clearly where you have `# Filter Client by registration date based on account type`.

**Comment by joncasdam** [comment]
True, in this case they are logically the same, but that is probably due to me trying to use the example model as it is.
So I'll update this with a different approach to make a better example.

**Comment by timgraham** [comment]
I could be wrong, but I still don't see why the new example is a good one. What does that query translate to in English? It looks to me like "get clients that are registered at least a year ago, regardless of newsletter status". If you have the same "then" clause, why use `Case`? What about something like this:
```
When(account_type=Client.GOLD, then=a_month_ago),
When(account_type=Client.PLATINUM, then=a_year_ago),
```
If my understanding is correct, this means get GOLD accounts registered more than a month ago and PLATINUM accounts registered more than a year ago.

**Comment by joncasdam** [comment]
I see but that isn't the thing with `When()` because it's similar to `filter()` as the docs states?

**Comment by timgraham** [comment]
I'm not sure what you mean.

**Comment by joncasdam** [comment]
@timgraham I go with your suggestion. I think now it's finally making more sense

**Comment by timgraham** [comment]
merged in b28c6ca7631e13ea29490bf51a367ab10198c74c, thanks!

**Pull request closed without merging**